Feat: replace plugin RPC with BTP

sharafian opened the pull request.

Commit:
feat: use CLP

Commit:
feat: add backwards compatible RPC class

Commit:
fix: syntax

Commit:
feat: add CLP support

Commit:
fix: remove multiple RPC URLs

Commit:
fix: emit requestId along with packet data

Commit:
fix: addressing comments

Commit:
feat: add auth token to RPC

Commit:
feat: send prefix in ws connection string

Commit:
feat: use CLP connection format from wiki

Commit:
fix: addressing comments

Commit:
fix: address comments

Commit:
fix: moved registerSideProtocolHandler into branch da-side-protocol

Commit:
fix: addressed comments

Commit:
feat: handle custom side-protocols

Commit:
fix: add socket properly on connect

Commit:
fix: throws an error if sendTransfer() received a CLP error (fixes #18)

Commit:
fix: call protocol handler with the correct object reference

Commit:
fix: change function signatures of prepare, fulfill, reject, message (fixes #22)

Commit:
fix: handle CLP rejection packets correctly

Commit:
refactor: CLP -> BTP

Commit:
feat: use new BTP error type

Commit:
refactor: move rejectionReason to protocolData

Commit:
refactor: fix lint errors

Commit:
chore: adds missing dependency ws

justmoon committed:
fix: update link to btp-packet

justmoon committed:
docs(readme): mockup for desired config format

Commit:
fix: protocolData must be array

Commit:
fix: parse the sub-protocol response correctly

justmoon committed:
feat: implement server and listener (wip)

Commit:
feat: add authentication on incoming/outgoing sockets

Commit:
feat: turn custom into a single json subprotocol

Commit:
test: fix tests with new features

Commit:
feat: always include/expect auth_username

Commit:
fix: use BTP URI format for client config

Commit:
fix: lint

Commit:
test: add test suggested by @dappelt

Commit:
fix: apply comments from @dappelt

Commit:
fix: address @sharafian review comments

Commit:
refactor: revert btpUri to server

Commit:
fix: authentication fixes

Commit:
feat: allow relying on plugin.addSocket

Commit:
test: add authentication tests

sharafian commented:
Should be ready to merge after review

sharafian commented:
sorry, accidentally hit close and comment

Commit:
style: lint

dappelt commented on test/infoSpec.js:
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
+        fulfillment: 'Ndr_HMuLPPl0idUlvAXFXBVQTFOizq-nXozej0KIA7k'
+      }, 100, []))
+
+      await new Promise((resolve) => {
```
You don't need to wrap this in a promise. `this.mockSocket.isDone()` makes sure all handlers registered via `.reply()` have been called.

dappelt commented on test/infoSpec.js:
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
+      this.newSocket = new MockSocket()
+      await this.plugin.addSocket(this.newSocket)
+    })
+
```
Check that all `MockSocket` handlers have been called by adding `afterEach`:

```js
  afterEach(async function () {
    assert(await this.mockSocket.isDone(), 'request handlers must have been called')
  })
```

sharafian replied:
added

dappelt commented on test/infoSpec.js:
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
+    it('should deny an authentication request with no "auth" protocol', async function () {
+      this.newSocket.emit('message', btpPacket.serializeMessage(100, []))
+
+      await new Promise((resolve) => {
```
No need for this Promise

dappelt commented on test/infoSpec.js:
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
No need for this Promise

dappelt commented on test/infoSpec.js:
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
No need for this Promise

dappelt commented on test/infoSpec.js:
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
No need for this Promise

dappelt commented on test/infoSpec.js:
```diff
@@ -129,6 +130,133 @@ describe('Info', () => {
+        data: Buffer.from('')
+      }]))
+
+      await new Promise((resolve) => {
```
No need for this Promise

michielbdejong commented:
It's a bit ugly that you need to specify both `opts.prefix` and `opts.info.prefix`, can you copy the second from the first?

Commit:
fix: refactor conflict in addSocket arguments

Commit:
fix: address comments from @dappelt and @michielbdejong

Commit:
feat: await auth

When calling addSocket as a client, the Promise returned by that
call doesn't resolve until auth has finished.
When calling addSocket as a server, the same should be true.

emschwartz commented on src/lib/client.js:
```diff
@@ -0,0 +1,44 @@
+    // Note that setting the parsedBtpUri.protocol does not work as expected,
+    // so removing the 'btp+' prefix from the full URL here:
+    assert(parsedBtpUri.toString().startsWith('btp+'), 'server uri must start with "btp+"')
+    this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
```
This seems like a bit of a janky way to either input `ws` or `wss`. It should be more explicit and throw an error if it's anything else

emschwartz replied:
Also shouldn't it take out the auth details?

dappelt replied:
Yes, it should take out auth details. That's done [here](https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/pull/21/files/c14320143b174a9c4262d320c0d804eecaa8fe5a#diff-976824b6113e95afacd333cbfab6d019R22).

emschwartz commented on test/channelSpec.js:
```diff
@@ -52,8 +57,9 @@ describe('makePaymentChannelPlugin', function () {
 
-    this.fulfillment = require('crypto').randomBytes(32)
-    this.transfer = {
+    this.fulfillment = crypto.randomBytes(32)
```
Probably better to just use a fixed value here

emschwartz reviewed (commented):
Overall looks okay. This module has a lot going on in it and it can be used in a lot of different ways so it might be worth breaking it up more

Commit:
feat: checkAuth function in plugin options

Commit:
fix: address comments from @emschwartz

sharafian commented:
Merging without tests passing because integration tests time out. This repo isn't used in the integration tests so it isn't an issue with the PR.
